File: Flink0105/src/main/scala/com/atguigu/day2/KeyedStreamExample.scala
package com.atguigu.day2

import org.apache.flink.streaming.api.scala._

object KeyedStreamExample {
  def main(args: Array[String]): Unit = {
    val env : StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    val stream : DataStream[SensorReading] = env
      .addSource(new SensorSource)
      .filter(_.id.equals("sensor_1"))

    // 泛型变成了两个，第二个泛型是key的类型
    val keyed : KeyedStream[SensorReading, String] = stream.keyBy(_.id)

    // 使用第三个字段来做滚动聚合，求每个传感器流上的最小温度值
    // 内部会保存一个最小值的状态变量，用来保存到来的温度的最小值
    // 滚动聚合以后，流的类型又变成了`DataStream`
    val min : DataStream[SensorReading] = keyed.min(2)
    min.print()

    // reduce也会保存一个状态变量
    val red : DataStream[SensorReading] = keyed.reduce((r1, r2) => SensorReading(r1.id, 0L, r1.temperature.min(r2.temperature)))
    red.print()

    env.execute()
  }
}

File: flink-java-code/src/main/java/com/atguigu/day2/KeyedStreamExample.java
package com.atguigu.day2;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.common.functions.ReduceFunction;
import org.apache.flink.api.common.typeinfo.TypeHint;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

public class KeyedStreamExample {
    public static void main(String[] args) throws Exception {
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStream<SensorReading> stream = env.addSource(new SensorSource());

        stream
                .map(r -> Tuple2.of(r.id, r.temperature))
                .returns(TypeInformation.of(new TypeHint<Tuple2<String, Double>>() { }))
                .filter(r -> r.f0.equals("sensor_1"))
                .keyBy(r -> r.f0)
                .reduce(new ReduceFunction<Tuple2<String, Double>>() {
                    @Override
                    public Tuple2<String, Double> reduce(Tuple2<String, Double> value1, Tuple2<String, Double> value2) throws Exception {
                        if (value1.f1 > value2.f1) {
                            return value1;
                        } else {
                            return value2;
                        }
                    }
                });

        stream
                .map(r -> new Tuple2<String, Double>(r.id, r.temperature))
                .returns(TypeInformation.of(new TypeHint<Tuple2<String, Double>>() { }))
                .filter(r -> r.f0.equals("sensor_1"))
                .keyBy(r -> r.f0)
                .reduce(new MyReduceFunction());

        stream
                .map(new MapFunction<SensorReading, Tuple2<String, Double>>() {
                    @Override
                    public Tuple2<String, Double> map(SensorReading value) throws Exception {
                        return new Tuple2<>(value.id, value.temperature);
                    }
                })
                .filter(r -> r.f0.equals("sensor_1"))
                .keyBy(r -> r.f0)
                .reduce(new MyReduceFunction())
                .print();

        env.execute();
    }

    public static class MyReduceFunction implements ReduceFunction<Tuple2<String, Double>> {
        @Override
        public Tuple2<String, Double> reduce(Tuple2<String, Double> value1, Tuple2<String, Double> value2) throws Exception {
            if (value1.f1 > value2.f1) {
                return value1;
            } else {
                return value2;
            }
        }
    }
}
